src/WLR/Types/Keyboard.hsc

```diff
@@ -0,0 +1,169 @@
+{-# LANGUAGE EmptyDataDeriving, PatternSynonyms #-}
+
+module WLR.Types.Keyboard where
+
+#define WLR_USE_UNSTABLE
+#include <wlr/types/wlr_keyboard.h>
+
+import Foreign (Word32, Storable(..))
+import Foreign.C.Types (CSize(..), CInt(..), CBool(..))
+import Foreign.C.String (CString)
+import Foreign.Ptr (Ptr)
+
+import WLR.Types.InputDevice (WLR_input_device)
+import WLR.Types.KeyboardGroup (WLR_keyboard_group)
+
+import WL.ServerCore (WL_signal)
+import WL.Keyboard (WL_keyboard_key_state)
+
+pattern WLR_LED_COUNT :: (Eq a, Num a) => a
+pattern WLR_LED_COUNT = #const WLR_LED_COUNT
+
+type WLR_keyboard_led = CInt
+-- enum wlr_keyboard_led
+--WLR_LED_NUM_LOCK = 1 << 0,
+pattern WLR_LED_NUM_LOCK :: (Eq a, Num a) => a
+pattern WLR_LED_NUM_LOCK = 1
+--WLR_LED_CAPS_LOCK = 1 << 1,
+pattern WLR_LED_CAPS_LOCK :: (Eq a, Num a) => a
+pattern WLR_LED_CAPS_LOCK = 2
+--WLR_LED_SCROLL_LOCK = 1 << 2,
+pattern WLR_LED_SCROLL_LOCK :: (Eq a, Num a) => a
+pattern WLR_LED_SCROLL_LOCK = 4
+
+pattern WLR_MODIFIER_COUNT :: (Eq a, Num a) => a
+pattern WLR_MODIFIER_COUNT = 8
+
+type Wlr_keyboard_modifier = CInt
+-- enum wlr_keyboard_modifier {
+-- WLR_MODIFIER_SHIFT = 1 << 0,
```

Comment:
I'm pretty sure these bit shifts are just declaring the value to be powers of two, instead of writing haskell code to calculate that I just wrote in the powers of 2 by hand

Reply:
Yep, that looks correct to me.